File: Brute Force/BJ 15686 치킨 배달.py
import math
import sys
from collections import deque
from collections import defaultdict
import itertools
import heapq
from bisect import bisect_left

sys.setrecursionlimit(10 ** 7)
input = sys.stdin.readline

n, m = map(int, input().split())
g = list(list(map(int, input().split())) for _ in range(n))

chick = []
home = []

for i in range(n):
    for j in range(n):
        if g[i][j] == 2:
            chick.append((i, j))
        elif g[i][j] == 1:
            home.append((i, j))

visit = set()


def calculate(visit):

    result = 0
    for h in home:

        result += min(list(abs(h[0] - v[0]) + abs(h[1] - v[1]) for v in visit))

    return result


result = 9e9

tmp = list([-1] * n for _ in range(n))


cl = len(chick)

# 조합을 백트래킹으로 구현하는 방법
def dfs(ori_idx, cnt):

    global result

    if cnt == m:
        result = min(result, calculate(visit))
        return

    for i in range(ori_idx + 1, cl):

        visit.add(chick[i])
        dfs(i, cnt + 1)
        visit.remove(chick[i])

for i in range(cl):

    visit.add(chick[i])
    dfs(i, 1)
    visit.remove(chick[i])


# 조합을 통한 풀이
# c = list(itertools.combinations(chick, m))
# result = 9e9
# for i in c:
#     result = min(result, calculate(i))

print(result)

File: DFS, BFS/java/BJ 15686 치킨 배달.java
import java.util.*;
import java.io.*;

public class Main {

    static int[][] g;
    static int n, m;

    static List<Node> chicken = new ArrayList();
    static List<Node> home = new ArrayList();
    static HashSet<Integer> visit = new HashSet<>();
    static int chickSize;
    static int result = Integer.MAX_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(bf.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        g = new int[n][n];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(bf.readLine());
            for (int j = 0; j < n; j++) {
                g[i][j] = Integer.parseInt(st.nextToken());

                if (g[i][j] == 1) {
                    home.add(new Node(i, j));
                } else if (g[i][j] == 2) {
                    chicken.add(new Node(i, j));
                }
            }
        }

        chickSize = chicken.size();

        for (int i = 0; i < chickSize; i++) {

            visit.add(i);
            dfs(i, 1);
            visit.remove(i);
        }

        System.out.println(result);

    }

    public static int calculate() {

        int result = 0;

        for (Node node : home) {

            int min = Integer.MAX_VALUE;

            for (Integer i : visit) {
                int tmp = Math.abs(node.y - chicken.get(i).y) + Math.abs(node.x - chicken.get(i).x);
                min = Math.min(min, tmp);
            }

            result += min;
        }

        return result;
    }

    public static void dfs(int oriIndex, int cnt) {

        if (cnt == m) {

            result = Math.min(result, calculate());
            return;
        }

        for (int i = oriIndex + 1; i < chickSize; i++) {

            visit.add(i);
            dfs(i, cnt + 1);
            visit.remove(i);
        }

    }


}

class Node {

    int y, x;

    public Node(int y, int x) {
        this.y = y;
        this.x = x;
    }

    @Override
    public int hashCode() {
        return Objects.hash(y, x);
    }

    @Override
    public boolean equals(Object obj) {

        if (this == obj)
            return true;

        if (!(obj instanceof Node)) {
            return false;
        }

        Node node = (Node) obj;
        return y == node.y && x == node.x;
    }
}

